{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Fisher's Exact Test to determine GO Term Enrichment of Probes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Boilerplate Import \n",
    "import sys\n",
    "sys.path.append('..')\n",
    "import os\n",
    "from itertools import chain\n",
    "from collections import defaultdict\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "from scripts import utils\n",
    "from scripts.analysis import build_loci"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "BASE = '../data'\n",
    "\n",
    "grouped_loci = utils.load(os.path.join(BASE,\"grouped_loci.pkl.gz\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "load obo file go-basic.obo\n",
      "go-basic.obo: fmt(1.2) rel(2016-07-16) 46,953 GO Terms\n"
     ]
    }
   ],
   "source": [
    "import gseapy as gp\n",
    "import goatools\n",
    "\n",
    "obodag = goatools.obo_parser.GODag('go-basic.obo')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def make_go2Gene_map(grouped_loci, ontology='MF'):\n",
    "    ontology_map = {'MF': 'molecular_function', 'BP': 'biological_process', 'CC': 'cellular_component'}\n",
    "    \n",
    "    out = defaultdict(list)    \n",
    "    for l in grouped_loci:\n",
    "        if 'go' in l.annotations:\n",
    "            for go in l.annotations['go']:\n",
    "                if obodag[go].namespace == ontology_map[ontology]:\n",
    "                    out[go].append(l.cluster_id)\n",
    "                    for parent in obodag[go].get_all_parents():\n",
    "                        if obodag[parent].namespace == ontology_map[ontology]:\n",
    "                            out[parent].append(l.cluster_id)\n",
    "                \n",
    "    return out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def filter_go2gene_map(go_locus):\n",
    "    \n",
    "    # Remove \"very broad\" gene sets. Arbitrary definition: gene sets that emcompass >50% of all IDs\n",
    "    all_ids = set(chain(*go_locus.values()))\n",
    "    go_locus = {key: value for (key, value) in go_locus.items() if len(value) / len(all_ids) <= 0.5}\n",
    "\n",
    "    # Remove terms with less than 5 members: changed from 10 to 5 becasue small #s of proteins compared\n",
    "    # to what you would find wiht genes\n",
    "    go_locus = {key: value for (key, value) in go_locus.items() if len(value) >= 10}\n",
    "\n",
    "    # Remove child terms with identical gene sets as their parents\n",
    "    to_remove = set()\n",
    "    for parent in go_locus.keys():\n",
    "        # If child term has exact same members as parent, remove\n",
    "        child_ids = [x.id for x in obodag[parent].children if x.id in go_locus.keys()]\n",
    "        for child in child_ids:\n",
    "            if go_locus[child] == go_locus[parent]:\n",
    "                to_remove.add(child)\n",
    "    go_locus = {key: value for (key, value) in go_locus.items() if key not in to_remove}\n",
    "\n",
    "    # Remove sibling terms with identical gene sets\n",
    "    to_remove = set()\n",
    "    for brother in go_locus.keys():\n",
    "        for parent in obodag[brother].parents:\n",
    "            siblings = set([y.id for y in parent.children])\n",
    "            siblings.remove(brother)\n",
    "            for sibling in siblings:\n",
    "                if sibling in go_locus.keys() and go_locus[brother] == go_locus[sibling]:\n",
    "                    to_remove.add(sibling)\n",
    "    go_locus = {key: value for (key, value) in go_locus.items() if key not in to_remove}\n",
    "\n",
    "    return go_locus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# pop is group1 + group2\n",
    "def enrichment_fish(go_in_group, nongo_in_group, go_in_other, nongo_in_other):\n",
    "    import scipy.stats as stats\n",
    "        \n",
    "    if go_in_group != 0:\n",
    "        go_in_group -= 1\n",
    "       \n",
    "    oddsr, pval = stats.fisher_exact([[go_in_group, go_in_other], [nongo_in_group, nongo_in_other]], alternative='two-sided')\n",
    "    \n",
    "    samplings = go_in_group + nongo_in_group\n",
    "    total_go = go_in_group + go_in_other\n",
    "    total = go_in_group + nongo_in_group + go_in_other + nongo_in_other\n",
    "\n",
    "    lower, expected = stats.hypergeom.interval(.99, M=total+1, n=total_go+1, N=samplings+1)\n",
    "    \n",
    "    return oddsr, pval, expected"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RAG_Enriched</th>\n",
       "      <th>RAG_Unenriched</th>\n",
       "      <th>RT_Enriched</th>\n",
       "      <th>RT_Unenriched</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>115324</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132035</th>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318873</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376417</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>407882</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       RAG_Enriched RAG_Unenriched RT_Enriched RT_Unenriched\n",
       "115324         True          False        True         False\n",
       "132035        False           True       False         False\n",
       "318873         True           True        True          True\n",
       "376417        False          False        True          True\n",
       "407882         True          False        True         False"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "group_df = pd.read_csv(os.path.join(BASE,'groups.csv'), index_col=0)\n",
    "group_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "mf_map = filter_go2gene_map(make_go2Gene_map(grouped_loci))\n",
    "bp_map = filter_go2gene_map(make_go2Gene_map(grouped_loci, ontology='BP'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def enrichment_df(group1, group2, go_map):\n",
    "    pv = dict()\n",
    "\n",
    "    all_annotations = set(go_map.keys())\n",
    "    \n",
    "    group_loci = set(group_df.query('{} and ~{}'.format(group1, group2)).index)\n",
    "    other_loci = set(group_df.query('~({} and ~{})'.format(group1, group2)).index)\n",
    "\n",
    "    for annotation in all_annotations:\n",
    "        go_in_group = len(set(go_map[annotation]) & group_loci)\n",
    "        nongo_in_group = len(group_loci) - go_in_group\n",
    "        go_in_other =  len(set(go_map[annotation]) & other_loci)\n",
    "        nongo_in_other = len(other_loci) - go_in_other\n",
    "\n",
    "        odds_r, p_val, expected = enrichment_fish(go_in_group, nongo_in_group, go_in_other, nongo_in_other)\n",
    "\n",
    "        pv[annotation] = [odds_r, p_val, go_in_group+go_in_other, go_in_group, expected]\n",
    "\n",
    "    res = (pd.DataFrame({obodag[k].name:v for k,v in pv.items()})\n",
    "             .T\n",
    "             .rename(columns={0:'odds_ratio', 1:'p_value', 2: 'total_annot', 3:'in_group', 4:'expected'})\n",
    "             .sort_values(by=['p_value']))\n",
    "    return res"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## RT - Enriched vs Uneneriched - MF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>odds_ratio</th>\n",
       "      <th>p_value</th>\n",
       "      <th>total_annot</th>\n",
       "      <th>in_group</th>\n",
       "      <th>expected</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>oxidoreductase activity, acting on other nitrogenous compounds as donors</th>\n",
       "      <td>22.729767</td>\n",
       "      <td>6.395769e-09</td>\n",
       "      <td>23.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>asparagine-tRNA ligase activity</th>\n",
       "      <td>14.381311</td>\n",
       "      <td>9.927729e-08</td>\n",
       "      <td>23.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cysteine-type peptidase activity</th>\n",
       "      <td>11.358025</td>\n",
       "      <td>1.481807e-07</td>\n",
       "      <td>25.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>14.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>urocanate hydratase activity</th>\n",
       "      <td>11.287466</td>\n",
       "      <td>2.609547e-04</td>\n",
       "      <td>13.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>alcohol dehydrogenase (NAD) activity</th>\n",
       "      <td>7.347440</td>\n",
       "      <td>1.373076e-04</td>\n",
       "      <td>18.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cysteine-type endopeptidase activity</th>\n",
       "      <td>6.777626</td>\n",
       "      <td>3.457332e-04</td>\n",
       "      <td>17.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>acetaldehyde dehydrogenase (acetylating) activity</th>\n",
       "      <td>6.777626</td>\n",
       "      <td>3.457332e-04</td>\n",
       "      <td>17.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>methionine-tRNA ligase activity</th>\n",
       "      <td>6.009977</td>\n",
       "      <td>5.133476e-03</td>\n",
       "      <td>12.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FMN binding</th>\n",
       "      <td>5.225979</td>\n",
       "      <td>4.619220e-06</td>\n",
       "      <td>34.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>18.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ligase activity, forming carbon-nitrogen bonds</th>\n",
       "      <td>4.388796</td>\n",
       "      <td>3.980853e-06</td>\n",
       "      <td>42.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>21.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ligase activity, forming carbon-oxygen bonds</th>\n",
       "      <td>4.208885</td>\n",
       "      <td>2.899121e-07</td>\n",
       "      <td>55.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>26.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aminoacyl-tRNA ligase activity</th>\n",
       "      <td>4.208885</td>\n",
       "      <td>2.899121e-07</td>\n",
       "      <td>55.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>26.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ligase activity</th>\n",
       "      <td>3.927544</td>\n",
       "      <td>1.013701e-11</td>\n",
       "      <td>106.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>45.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oxidoreductase activity, acting on the aldehyde or oxo group of donors, NAD or NADP as acceptor</th>\n",
       "      <td>2.733317</td>\n",
       "      <td>3.347309e-09</td>\n",
       "      <td>149.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>61.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hydrolase activity, acting on carbon-nitrogen (but not peptide) bonds</th>\n",
       "      <td>2.699669</td>\n",
       "      <td>1.060250e-03</td>\n",
       "      <td>47.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>23.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>phosphotransferase activity, paired acceptors</th>\n",
       "      <td>2.121503</td>\n",
       "      <td>3.923154e-07</td>\n",
       "      <td>200.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>78.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>peptidase activity</th>\n",
       "      <td>1.996151</td>\n",
       "      <td>5.828818e-04</td>\n",
       "      <td>106.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>45.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oxidoreductase activity, acting on the CH-NH2 group of donors</th>\n",
       "      <td>1.889868</td>\n",
       "      <td>1.077896e-03</td>\n",
       "      <td>118.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>49.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transferase activity, transferring acyl groups</th>\n",
       "      <td>1.800678</td>\n",
       "      <td>6.705602e-03</td>\n",
       "      <td>96.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>41.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>carbon-oxygen lyase activity</th>\n",
       "      <td>1.768311</td>\n",
       "      <td>3.353492e-03</td>\n",
       "      <td>120.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>50.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oxidoreductase activity, acting on the CH-NH2 group of donors, NAD or NADP as acceptor</th>\n",
       "      <td>1.714099</td>\n",
       "      <td>7.851519e-03</td>\n",
       "      <td>108.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>46.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hydro-lyase activity</th>\n",
       "      <td>1.697918</td>\n",
       "      <td>6.056570e-03</td>\n",
       "      <td>118.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>49.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oxidoreductase activity</th>\n",
       "      <td>1.644580</td>\n",
       "      <td>9.651185e-11</td>\n",
       "      <td>928.0</td>\n",
       "      <td>370.0</td>\n",
       "      <td>319.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    odds_ratio       p_value  \\\n",
       "oxidoreductase activity, acting on other nitrog...   22.729767  6.395769e-09   \n",
       "asparagine-tRNA ligase activity                      14.381311  9.927729e-08   \n",
       "cysteine-type peptidase activity                     11.358025  1.481807e-07   \n",
       "urocanate hydratase activity                         11.287466  2.609547e-04   \n",
       "alcohol dehydrogenase (NAD) activity                  7.347440  1.373076e-04   \n",
       "cysteine-type endopeptidase activity                  6.777626  3.457332e-04   \n",
       "acetaldehyde dehydrogenase (acetylating) activity     6.777626  3.457332e-04   \n",
       "methionine-tRNA ligase activity                       6.009977  5.133476e-03   \n",
       "FMN binding                                           5.225979  4.619220e-06   \n",
       "ligase activity, forming carbon-nitrogen bonds        4.388796  3.980853e-06   \n",
       "ligase activity, forming carbon-oxygen bonds          4.208885  2.899121e-07   \n",
       "aminoacyl-tRNA ligase activity                        4.208885  2.899121e-07   \n",
       "ligase activity                                       3.927544  1.013701e-11   \n",
       "oxidoreductase activity, acting on the aldehyde...    2.733317  3.347309e-09   \n",
       "hydrolase activity, acting on carbon-nitrogen (...    2.699669  1.060250e-03   \n",
       "phosphotransferase activity, paired acceptors         2.121503  3.923154e-07   \n",
       "peptidase activity                                    1.996151  5.828818e-04   \n",
       "oxidoreductase activity, acting on the CH-NH2 g...    1.889868  1.077896e-03   \n",
       "transferase activity, transferring acyl groups        1.800678  6.705602e-03   \n",
       "carbon-oxygen lyase activity                          1.768311  3.353492e-03   \n",
       "oxidoreductase activity, acting on the CH-NH2 g...    1.714099  7.851519e-03   \n",
       "hydro-lyase activity                                  1.697918  6.056570e-03   \n",
       "oxidoreductase activity                               1.644580  9.651185e-11   \n",
       "\n",
       "                                                    total_annot  in_group  \\\n",
       "oxidoreductase activity, acting on other nitrog...         23.0      21.0   \n",
       "asparagine-tRNA ligase activity                            23.0      20.0   \n",
       "cysteine-type peptidase activity                           25.0      21.0   \n",
       "urocanate hydratase activity                               13.0      11.0   \n",
       "alcohol dehydrogenase (NAD) activity                       18.0      14.0   \n",
       "cysteine-type endopeptidase activity                       17.0      13.0   \n",
       "acetaldehyde dehydrogenase (acetylating) activity          17.0      13.0   \n",
       "methionine-tRNA ligase activity                            12.0       9.0   \n",
       "FMN binding                                                34.0      24.0   \n",
       "ligase activity, forming carbon-nitrogen bonds             42.0      28.0   \n",
       "ligase activity, forming carbon-oxygen bonds               55.0      36.0   \n",
       "aminoacyl-tRNA ligase activity                             55.0      36.0   \n",
       "ligase activity                                           106.0      67.0   \n",
       "oxidoreductase activity, acting on the aldehyde...        149.0      81.0   \n",
       "hydrolase activity, acting on carbon-nitrogen (...         47.0      26.0   \n",
       "phosphotransferase activity, paired acceptors             200.0      96.0   \n",
       "peptidase activity                                        106.0      50.0   \n",
       "oxidoreductase activity, acting on the CH-NH2 g...        118.0      54.0   \n",
       "transferase activity, transferring acyl groups             96.0      43.0   \n",
       "carbon-oxygen lyase activity                              120.0      53.0   \n",
       "oxidoreductase activity, acting on the CH-NH2 g...        108.0      47.0   \n",
       "hydro-lyase activity                                      118.0      51.0   \n",
       "oxidoreductase activity                                   928.0     370.0   \n",
       "\n",
       "                                                    expected  \n",
       "oxidoreductase activity, acting on other nitrog...      13.0  \n",
       "asparagine-tRNA ligase activity                         13.0  \n",
       "cysteine-type peptidase activity                        14.0  \n",
       "urocanate hydratase activity                             8.0  \n",
       "alcohol dehydrogenase (NAD) activity                    11.0  \n",
       "cysteine-type endopeptidase activity                    10.0  \n",
       "acetaldehyde dehydrogenase (acetylating) activity       10.0  \n",
       "methionine-tRNA ligase activity                          8.0  \n",
       "FMN binding                                             18.0  \n",
       "ligase activity, forming carbon-nitrogen bonds          21.0  \n",
       "ligase activity, forming carbon-oxygen bonds            26.0  \n",
       "aminoacyl-tRNA ligase activity                          26.0  \n",
       "ligase activity                                         45.0  \n",
       "oxidoreductase activity, acting on the aldehyde...      61.0  \n",
       "hydrolase activity, acting on carbon-nitrogen (...      23.0  \n",
       "phosphotransferase activity, paired acceptors           78.0  \n",
       "peptidase activity                                      45.0  \n",
       "oxidoreductase activity, acting on the CH-NH2 g...      49.0  \n",
       "transferase activity, transferring acyl groups          41.0  \n",
       "carbon-oxygen lyase activity                            50.0  \n",
       "oxidoreductase activity, acting on the CH-NH2 g...      46.0  \n",
       "hydro-lyase activity                                    49.0  \n",
       "oxidoreductase activity                                319.0  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(enrichment_df('RT_Enriched', 'RT_Unenriched', mf_map)\n",
    " .query('odds_ratio > 1 and p_value < 0.01')\n",
    " .sort_values('odds_ratio', ascending=False))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## RT - Enriched vs Uneneriched - MF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>odds_ratio</th>\n",
       "      <th>p_value</th>\n",
       "      <th>total_annot</th>\n",
       "      <th>in_group</th>\n",
       "      <th>expected</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>serine family amino acid catabolic process</th>\n",
       "      <td>15.774983</td>\n",
       "      <td>1.533337e-03</td>\n",
       "      <td>9.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>asparaginyl-tRNA aminoacylation</th>\n",
       "      <td>14.381311</td>\n",
       "      <td>9.927729e-08</td>\n",
       "      <td>23.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>carbon utilization</th>\n",
       "      <td>6.777626</td>\n",
       "      <td>3.457332e-04</td>\n",
       "      <td>17.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>methionyl-tRNA aminoacylation</th>\n",
       "      <td>6.009977</td>\n",
       "      <td>5.133476e-03</td>\n",
       "      <td>12.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ncRNA metabolic process</th>\n",
       "      <td>4.332141</td>\n",
       "      <td>1.458621e-07</td>\n",
       "      <td>56.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>26.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tRNA aminoacylation for protein translation</th>\n",
       "      <td>4.208885</td>\n",
       "      <td>2.899121e-07</td>\n",
       "      <td>55.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>26.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>amino acid activation</th>\n",
       "      <td>4.208885</td>\n",
       "      <td>2.899121e-07</td>\n",
       "      <td>55.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>26.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>branched-chain amino acid metabolic process</th>\n",
       "      <td>3.585561</td>\n",
       "      <td>5.821360e-04</td>\n",
       "      <td>32.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>17.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>organic acid catabolic process</th>\n",
       "      <td>3.161475</td>\n",
       "      <td>6.346572e-03</td>\n",
       "      <td>25.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>14.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein refolding</th>\n",
       "      <td>2.768468</td>\n",
       "      <td>1.619826e-03</td>\n",
       "      <td>41.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>20.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cellular amino acid metabolic process</th>\n",
       "      <td>2.033452</td>\n",
       "      <td>3.020062e-10</td>\n",
       "      <td>359.0</td>\n",
       "      <td>166.0</td>\n",
       "      <td>133.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>proteolysis</th>\n",
       "      <td>1.984031</td>\n",
       "      <td>9.765979e-04</td>\n",
       "      <td>102.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>44.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oxidation-reduction process</th>\n",
       "      <td>1.491910</td>\n",
       "      <td>2.153628e-07</td>\n",
       "      <td>933.0</td>\n",
       "      <td>355.0</td>\n",
       "      <td>321.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>organic acid metabolic process</th>\n",
       "      <td>1.477733</td>\n",
       "      <td>5.676488e-07</td>\n",
       "      <td>907.0</td>\n",
       "      <td>344.0</td>\n",
       "      <td>312.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>small molecule metabolic process</th>\n",
       "      <td>1.428382</td>\n",
       "      <td>3.513252e-07</td>\n",
       "      <td>1255.0</td>\n",
       "      <td>460.0</td>\n",
       "      <td>423.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>carboxylic acid metabolic process</th>\n",
       "      <td>1.420662</td>\n",
       "      <td>9.808493e-06</td>\n",
       "      <td>875.0</td>\n",
       "      <td>326.0</td>\n",
       "      <td>302.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oxoacid metabolic process</th>\n",
       "      <td>1.414483</td>\n",
       "      <td>1.207537e-05</td>\n",
       "      <td>877.0</td>\n",
       "      <td>326.0</td>\n",
       "      <td>303.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             odds_ratio       p_value  \\\n",
       "serine family amino acid catabolic process    15.774983  1.533337e-03   \n",
       "asparaginyl-tRNA aminoacylation               14.381311  9.927729e-08   \n",
       "carbon utilization                             6.777626  3.457332e-04   \n",
       "methionyl-tRNA aminoacylation                  6.009977  5.133476e-03   \n",
       "ncRNA metabolic process                        4.332141  1.458621e-07   \n",
       "tRNA aminoacylation for protein translation    4.208885  2.899121e-07   \n",
       "amino acid activation                          4.208885  2.899121e-07   \n",
       "branched-chain amino acid metabolic process    3.585561  5.821360e-04   \n",
       "organic acid catabolic process                 3.161475  6.346572e-03   \n",
       "protein refolding                              2.768468  1.619826e-03   \n",
       "cellular amino acid metabolic process          2.033452  3.020062e-10   \n",
       "proteolysis                                    1.984031  9.765979e-04   \n",
       "oxidation-reduction process                    1.491910  2.153628e-07   \n",
       "organic acid metabolic process                 1.477733  5.676488e-07   \n",
       "small molecule metabolic process               1.428382  3.513252e-07   \n",
       "carboxylic acid metabolic process              1.420662  9.808493e-06   \n",
       "oxoacid metabolic process                      1.414483  1.207537e-05   \n",
       "\n",
       "                                             total_annot  in_group  expected  \n",
       "serine family amino acid catabolic process           9.0       8.0       7.0  \n",
       "asparaginyl-tRNA aminoacylation                     23.0      20.0      13.0  \n",
       "carbon utilization                                  17.0      13.0      10.0  \n",
       "methionyl-tRNA aminoacylation                       12.0       9.0       8.0  \n",
       "ncRNA metabolic process                             56.0      37.0      26.0  \n",
       "tRNA aminoacylation for protein translation         55.0      36.0      26.0  \n",
       "amino acid activation                               55.0      36.0      26.0  \n",
       "branched-chain amino acid metabolic process         32.0      20.0      17.0  \n",
       "organic acid catabolic process                      25.0      15.0      14.0  \n",
       "protein refolding                                   41.0      23.0      20.0  \n",
       "cellular amino acid metabolic process              359.0     166.0     133.0  \n",
       "proteolysis                                        102.0      48.0      44.0  \n",
       "oxidation-reduction process                        933.0     355.0     321.0  \n",
       "organic acid metabolic process                     907.0     344.0     312.0  \n",
       "small molecule metabolic process                  1255.0     460.0     423.0  \n",
       "carboxylic acid metabolic process                  875.0     326.0     302.0  \n",
       "oxoacid metabolic process                          877.0     326.0     303.0  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(enrichment_df('RT_Enriched', 'RT_Unenriched', bp_map)\n",
    " .query('odds_ratio > 1 and p_value < 0.01')\n",
    " .sort_values('odds_ratio', ascending=False))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## RT - Unenriched vs Enriched - MF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>odds_ratio</th>\n",
       "      <th>p_value</th>\n",
       "      <th>total_annot</th>\n",
       "      <th>in_group</th>\n",
       "      <th>expected</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>phosphoglycerate mutase activity</th>\n",
       "      <td>13.085006</td>\n",
       "      <td>2.973009e-12</td>\n",
       "      <td>45.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>26.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein-N(PI)-phosphohistidine-sugar phosphotransferase activity</th>\n",
       "      <td>11.591440</td>\n",
       "      <td>5.998320e-05</td>\n",
       "      <td>17.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>intramolecular transferase activity, phosphotransferases</th>\n",
       "      <td>8.252778</td>\n",
       "      <td>1.452016e-12</td>\n",
       "      <td>60.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>32.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>methionine adenosyltransferase activity</th>\n",
       "      <td>7.111075</td>\n",
       "      <td>1.122956e-15</td>\n",
       "      <td>84.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>43.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>structural molecule activity</th>\n",
       "      <td>6.837471</td>\n",
       "      <td>1.845553e-54</td>\n",
       "      <td>327.0</td>\n",
       "      <td>257.0</td>\n",
       "      <td>146.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA polymerase activity</th>\n",
       "      <td>6.766247</td>\n",
       "      <td>8.142000e-25</td>\n",
       "      <td>143.0</td>\n",
       "      <td>114.0</td>\n",
       "      <td>69.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transferase activity, transferring alkyl or aryl (other than methyl) groups</th>\n",
       "      <td>6.690520</td>\n",
       "      <td>2.883961e-15</td>\n",
       "      <td>85.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>44.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DNA binding</th>\n",
       "      <td>6.504838</td>\n",
       "      <td>9.505842e-25</td>\n",
       "      <td>148.0</td>\n",
       "      <td>117.0</td>\n",
       "      <td>71.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C-acetyltransferase activity</th>\n",
       "      <td>6.059563</td>\n",
       "      <td>3.557904e-03</td>\n",
       "      <td>15.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>structural constituent of ribosome</th>\n",
       "      <td>5.815262</td>\n",
       "      <td>6.828591e-27</td>\n",
       "      <td>178.0</td>\n",
       "      <td>137.0</td>\n",
       "      <td>84.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>formate C-acetyltransferase activity</th>\n",
       "      <td>5.505639</td>\n",
       "      <td>6.876722e-03</td>\n",
       "      <td>14.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>carbohydrate transporter activity</th>\n",
       "      <td>4.631907</td>\n",
       "      <td>1.722973e-03</td>\n",
       "      <td>20.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C-acyltransferase activity</th>\n",
       "      <td>4.543147</td>\n",
       "      <td>6.424926e-03</td>\n",
       "      <td>16.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nucleic acid binding</th>\n",
       "      <td>4.296453</td>\n",
       "      <td>9.753064e-44</td>\n",
       "      <td>425.0</td>\n",
       "      <td>296.0</td>\n",
       "      <td>186.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA binding</th>\n",
       "      <td>3.606458</td>\n",
       "      <td>1.993056e-21</td>\n",
       "      <td>245.0</td>\n",
       "      <td>165.0</td>\n",
       "      <td>112.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nucleotidyltransferase activity</th>\n",
       "      <td>3.347725</td>\n",
       "      <td>8.255304e-18</td>\n",
       "      <td>223.0</td>\n",
       "      <td>147.0</td>\n",
       "      <td>103.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATPase activity, coupled to transmembrane movement of ions, rotational mechanism</th>\n",
       "      <td>2.963207</td>\n",
       "      <td>1.279814e-03</td>\n",
       "      <td>40.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>23.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>proton-transporting ATP synthase activity, rotational mechanism</th>\n",
       "      <td>2.963207</td>\n",
       "      <td>1.279814e-03</td>\n",
       "      <td>40.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>23.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>substrate-specific transmembrane transporter activity</th>\n",
       "      <td>2.892773</td>\n",
       "      <td>1.825501e-05</td>\n",
       "      <td>72.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>38.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>phosphotransferase activity, carboxyl group as acceptor</th>\n",
       "      <td>2.890183</td>\n",
       "      <td>2.400003e-04</td>\n",
       "      <td>53.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>29.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>substrate-specific transporter activity</th>\n",
       "      <td>2.848184</td>\n",
       "      <td>1.409212e-05</td>\n",
       "      <td>74.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>39.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hydrogen ion transmembrane transporter activity</th>\n",
       "      <td>2.696209</td>\n",
       "      <td>2.001962e-03</td>\n",
       "      <td>43.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>25.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>intramolecular transferase activity</th>\n",
       "      <td>2.667406</td>\n",
       "      <td>1.672724e-05</td>\n",
       "      <td>84.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>43.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>active ion transmembrane transporter activity</th>\n",
       "      <td>2.536752</td>\n",
       "      <td>3.876395e-03</td>\n",
       "      <td>44.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>25.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>inorganic cation transmembrane transporter activity</th>\n",
       "      <td>2.488522</td>\n",
       "      <td>2.961660e-03</td>\n",
       "      <td>46.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>26.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cation transmembrane transporter activity</th>\n",
       "      <td>2.445408</td>\n",
       "      <td>3.569098e-03</td>\n",
       "      <td>48.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>27.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>phosphoglycerate kinase activity</th>\n",
       "      <td>2.356751</td>\n",
       "      <td>5.337309e-03</td>\n",
       "      <td>47.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>26.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATPase activity, coupled</th>\n",
       "      <td>2.322353</td>\n",
       "      <td>4.189826e-03</td>\n",
       "      <td>49.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>27.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>thiamine pyrophosphate binding</th>\n",
       "      <td>2.257827</td>\n",
       "      <td>1.242706e-03</td>\n",
       "      <td>67.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>primary active transmembrane transporter activity</th>\n",
       "      <td>2.238158</td>\n",
       "      <td>6.366444e-03</td>\n",
       "      <td>48.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>27.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ion transmembrane transporter activity</th>\n",
       "      <td>2.211018</td>\n",
       "      <td>7.118052e-03</td>\n",
       "      <td>50.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>28.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTPase activity</th>\n",
       "      <td>1.640024</td>\n",
       "      <td>6.673480e-03</td>\n",
       "      <td>126.0</td>\n",
       "      <td>63.0</td>\n",
       "      <td>62.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>isomerase activity</th>\n",
       "      <td>1.575416</td>\n",
       "      <td>2.419213e-03</td>\n",
       "      <td>195.0</td>\n",
       "      <td>95.0</td>\n",
       "      <td>91.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transferase activity</th>\n",
       "      <td>1.411449</td>\n",
       "      <td>4.067492e-06</td>\n",
       "      <td>927.0</td>\n",
       "      <td>413.0</td>\n",
       "      <td>385.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transferase activity, transferring phosphorus-containing groups</th>\n",
       "      <td>1.276681</td>\n",
       "      <td>6.484501e-03</td>\n",
       "      <td>590.0</td>\n",
       "      <td>254.0</td>\n",
       "      <td>252.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    odds_ratio       p_value  \\\n",
       "phosphoglycerate mutase activity                     13.085006  2.973009e-12   \n",
       "protein-N(PI)-phosphohistidine-sugar phosphotra...   11.591440  5.998320e-05   \n",
       "intramolecular transferase activity, phosphotra...    8.252778  1.452016e-12   \n",
       "methionine adenosyltransferase activity               7.111075  1.122956e-15   \n",
       "structural molecule activity                          6.837471  1.845553e-54   \n",
       "RNA polymerase activity                               6.766247  8.142000e-25   \n",
       "transferase activity, transferring alkyl or ary...    6.690520  2.883961e-15   \n",
       "DNA binding                                           6.504838  9.505842e-25   \n",
       "C-acetyltransferase activity                          6.059563  3.557904e-03   \n",
       "structural constituent of ribosome                    5.815262  6.828591e-27   \n",
       "formate C-acetyltransferase activity                  5.505639  6.876722e-03   \n",
       "carbohydrate transporter activity                     4.631907  1.722973e-03   \n",
       "C-acyltransferase activity                            4.543147  6.424926e-03   \n",
       "nucleic acid binding                                  4.296453  9.753064e-44   \n",
       "RNA binding                                           3.606458  1.993056e-21   \n",
       "nucleotidyltransferase activity                       3.347725  8.255304e-18   \n",
       "ATPase activity, coupled to transmembrane movem...    2.963207  1.279814e-03   \n",
       "proton-transporting ATP synthase activity, rota...    2.963207  1.279814e-03   \n",
       "substrate-specific transmembrane transporter ac...    2.892773  1.825501e-05   \n",
       "phosphotransferase activity, carboxyl group as ...    2.890183  2.400003e-04   \n",
       "substrate-specific transporter activity               2.848184  1.409212e-05   \n",
       "hydrogen ion transmembrane transporter activity       2.696209  2.001962e-03   \n",
       "intramolecular transferase activity                   2.667406  1.672724e-05   \n",
       "active ion transmembrane transporter activity         2.536752  3.876395e-03   \n",
       "inorganic cation transmembrane transporter acti...    2.488522  2.961660e-03   \n",
       "cation transmembrane transporter activity             2.445408  3.569098e-03   \n",
       "phosphoglycerate kinase activity                      2.356751  5.337309e-03   \n",
       "ATPase activity, coupled                              2.322353  4.189826e-03   \n",
       "thiamine pyrophosphate binding                        2.257827  1.242706e-03   \n",
       "primary active transmembrane transporter activity     2.238158  6.366444e-03   \n",
       "ion transmembrane transporter activity                2.211018  7.118052e-03   \n",
       "GTPase activity                                       1.640024  6.673480e-03   \n",
       "isomerase activity                                    1.575416  2.419213e-03   \n",
       "transferase activity                                  1.411449  4.067492e-06   \n",
       "transferase activity, transferring phosphorus-c...    1.276681  6.484501e-03   \n",
       "\n",
       "                                                    total_annot  in_group  \\\n",
       "phosphoglycerate mutase activity                           45.0      40.0   \n",
       "protein-N(PI)-phosphohistidine-sugar phosphotra...         17.0      15.0   \n",
       "intramolecular transferase activity, phosphotra...         60.0      50.0   \n",
       "methionine adenosyltransferase activity                    84.0      68.0   \n",
       "structural molecule activity                              327.0     257.0   \n",
       "RNA polymerase activity                                   143.0     114.0   \n",
       "transferase activity, transferring alkyl or ary...         85.0      68.0   \n",
       "DNA binding                                               148.0     117.0   \n",
       "C-acetyltransferase activity                               15.0      12.0   \n",
       "structural constituent of ribosome                        178.0     137.0   \n",
       "formate C-acetyltransferase activity                       14.0      11.0   \n",
       "carbohydrate transporter activity                          20.0      15.0   \n",
       "C-acyltransferase activity                                 16.0      12.0   \n",
       "nucleic acid binding                                      425.0     296.0   \n",
       "RNA binding                                               245.0     165.0   \n",
       "nucleotidyltransferase activity                           223.0     147.0   \n",
       "ATPase activity, coupled to transmembrane movem...         40.0      26.0   \n",
       "proton-transporting ATP synthase activity, rota...         40.0      26.0   \n",
       "substrate-specific transmembrane transporter ac...         72.0      46.0   \n",
       "phosphotransferase activity, carboxyl group as ...         53.0      34.0   \n",
       "substrate-specific transporter activity                    74.0      47.0   \n",
       "hydrogen ion transmembrane transporter activity            43.0      27.0   \n",
       "intramolecular transferase activity                        84.0      52.0   \n",
       "active ion transmembrane transporter activity              44.0      27.0   \n",
       "inorganic cation transmembrane transporter acti...         46.0      28.0   \n",
       "cation transmembrane transporter activity                  48.0      29.0   \n",
       "phosphoglycerate kinase activity                           47.0      28.0   \n",
       "ATPase activity, coupled                                   49.0      29.0   \n",
       "thiamine pyrophosphate binding                             67.0      39.0   \n",
       "primary active transmembrane transporter activity          48.0      28.0   \n",
       "ion transmembrane transporter activity                     50.0      29.0   \n",
       "GTPase activity                                           126.0      63.0   \n",
       "isomerase activity                                        195.0      95.0   \n",
       "transferase activity                                      927.0     413.0   \n",
       "transferase activity, transferring phosphorus-c...        590.0     254.0   \n",
       "\n",
       "                                                    expected  \n",
       "phosphoglycerate mutase activity                        26.0  \n",
       "protein-N(PI)-phosphohistidine-sugar phosphotra...      12.0  \n",
       "intramolecular transferase activity, phosphotra...      32.0  \n",
       "methionine adenosyltransferase activity                 43.0  \n",
       "structural molecule activity                           146.0  \n",
       "RNA polymerase activity                                 69.0  \n",
       "transferase activity, transferring alkyl or ary...      44.0  \n",
       "DNA binding                                             71.0  \n",
       "C-acetyltransferase activity                            11.0  \n",
       "structural constituent of ribosome                      84.0  \n",
       "formate C-acetyltransferase activity                    10.0  \n",
       "carbohydrate transporter activity                       13.0  \n",
       "C-acyltransferase activity                              11.0  \n",
       "nucleic acid binding                                   186.0  \n",
       "RNA binding                                            112.0  \n",
       "nucleotidyltransferase activity                        103.0  \n",
       "ATPase activity, coupled to transmembrane movem...      23.0  \n",
       "proton-transporting ATP synthase activity, rota...      23.0  \n",
       "substrate-specific transmembrane transporter ac...      38.0  \n",
       "phosphotransferase activity, carboxyl group as ...      29.0  \n",
       "substrate-specific transporter activity                 39.0  \n",
       "hydrogen ion transmembrane transporter activity         25.0  \n",
       "intramolecular transferase activity                     43.0  \n",
       "active ion transmembrane transporter activity           25.0  \n",
       "inorganic cation transmembrane transporter acti...      26.0  \n",
       "cation transmembrane transporter activity               27.0  \n",
       "phosphoglycerate kinase activity                        26.0  \n",
       "ATPase activity, coupled                                27.0  \n",
       "thiamine pyrophosphate binding                          36.0  \n",
       "primary active transmembrane transporter activity       27.0  \n",
       "ion transmembrane transporter activity                  28.0  \n",
       "GTPase activity                                         62.0  \n",
       "isomerase activity                                      91.0  \n",
       "transferase activity                                   385.0  \n",
       "transferase activity, transferring phosphorus-c...     252.0  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(enrichment_df('RT_Unenriched', 'RT_Enriched', mf_map)\n",
    " .query('odds_ratio > 1 and p_value < 0.01')\n",
    " .sort_values('odds_ratio', ascending=False))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## RT - Unenriched vs Enriched - BP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>odds_ratio</th>\n",
       "      <th>p_value</th>\n",
       "      <th>total_annot</th>\n",
       "      <th>in_group</th>\n",
       "      <th>expected</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>movement of cell or subcellular component</th>\n",
       "      <td>99.007941</td>\n",
       "      <td>5.477514e-46</td>\n",
       "      <td>116.0</td>\n",
       "      <td>114.0</td>\n",
       "      <td>57.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>locomotion</th>\n",
       "      <td>99.007941</td>\n",
       "      <td>5.477514e-46</td>\n",
       "      <td>116.0</td>\n",
       "      <td>114.0</td>\n",
       "      <td>57.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S-adenosylmethionine metabolic process</th>\n",
       "      <td>7.111075</td>\n",
       "      <td>1.122956e-15</td>\n",
       "      <td>84.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>43.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA biosynthetic process</th>\n",
       "      <td>6.894115</td>\n",
       "      <td>2.009358e-25</td>\n",
       "      <td>145.0</td>\n",
       "      <td>116.0</td>\n",
       "      <td>70.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nucleic acid-templated transcription</th>\n",
       "      <td>6.766247</td>\n",
       "      <td>8.142000e-25</td>\n",
       "      <td>143.0</td>\n",
       "      <td>114.0</td>\n",
       "      <td>69.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sulfur compound biosynthetic process</th>\n",
       "      <td>6.690520</td>\n",
       "      <td>2.883961e-15</td>\n",
       "      <td>85.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>44.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>amide biosynthetic process</th>\n",
       "      <td>6.037490</td>\n",
       "      <td>3.428364e-29</td>\n",
       "      <td>187.0</td>\n",
       "      <td>145.0</td>\n",
       "      <td>88.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>peptide metabolic process</th>\n",
       "      <td>5.895077</td>\n",
       "      <td>1.094698e-28</td>\n",
       "      <td>188.0</td>\n",
       "      <td>145.0</td>\n",
       "      <td>88.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cellular amide metabolic process</th>\n",
       "      <td>5.846131</td>\n",
       "      <td>5.978251e-29</td>\n",
       "      <td>191.0</td>\n",
       "      <td>147.0</td>\n",
       "      <td>89.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>macromolecule biosynthetic process</th>\n",
       "      <td>4.639342</td>\n",
       "      <td>1.728650e-48</td>\n",
       "      <td>432.0</td>\n",
       "      <td>307.0</td>\n",
       "      <td>188.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cellular macromolecule biosynthetic process</th>\n",
       "      <td>4.602911</td>\n",
       "      <td>6.335099e-48</td>\n",
       "      <td>430.0</td>\n",
       "      <td>305.0</td>\n",
       "      <td>188.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coenzyme biosynthetic process</th>\n",
       "      <td>4.270630</td>\n",
       "      <td>3.248741e-12</td>\n",
       "      <td>103.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>52.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cellular nitrogen compound biosynthetic process</th>\n",
       "      <td>3.950762</td>\n",
       "      <td>1.881661e-42</td>\n",
       "      <td>462.0</td>\n",
       "      <td>313.0</td>\n",
       "      <td>200.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>regulation of gene expression</th>\n",
       "      <td>3.858625</td>\n",
       "      <td>4.460849e-03</td>\n",
       "      <td>21.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>14.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nucleobase-containing compound biosynthetic process</th>\n",
       "      <td>3.323436</td>\n",
       "      <td>1.538432e-18</td>\n",
       "      <td>236.0</td>\n",
       "      <td>155.0</td>\n",
       "      <td>108.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sulfur compound metabolic process</th>\n",
       "      <td>3.202878</td>\n",
       "      <td>3.417680e-09</td>\n",
       "      <td>111.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>55.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>phosphoenolpyruvate-dependent sugar phosphotransferase system</th>\n",
       "      <td>3.179097</td>\n",
       "      <td>1.243402e-03</td>\n",
       "      <td>36.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>21.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cellular biosynthetic process</th>\n",
       "      <td>2.990730</td>\n",
       "      <td>1.842103e-41</td>\n",
       "      <td>742.0</td>\n",
       "      <td>448.0</td>\n",
       "      <td>312.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cellular protein metabolic process</th>\n",
       "      <td>2.934089</td>\n",
       "      <td>5.018869e-17</td>\n",
       "      <td>266.0</td>\n",
       "      <td>167.0</td>\n",
       "      <td>121.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA metabolic process</th>\n",
       "      <td>2.926322</td>\n",
       "      <td>5.382106e-15</td>\n",
       "      <td>232.0</td>\n",
       "      <td>146.0</td>\n",
       "      <td>106.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cofactor biosynthetic process</th>\n",
       "      <td>2.880466</td>\n",
       "      <td>2.447821e-08</td>\n",
       "      <td>120.0</td>\n",
       "      <td>76.0</td>\n",
       "      <td>59.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>energy coupled proton transport, down electrochemical gradient</th>\n",
       "      <td>2.876926</td>\n",
       "      <td>1.012332e-03</td>\n",
       "      <td>42.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>24.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aromatic compound biosynthetic process</th>\n",
       "      <td>2.788336</td>\n",
       "      <td>1.540568e-15</td>\n",
       "      <td>266.0</td>\n",
       "      <td>164.0</td>\n",
       "      <td>121.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nucleic acid metabolic process</th>\n",
       "      <td>2.719766</td>\n",
       "      <td>9.230484e-14</td>\n",
       "      <td>240.0</td>\n",
       "      <td>147.0</td>\n",
       "      <td>110.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cellular macromolecule metabolic process</th>\n",
       "      <td>2.718431</td>\n",
       "      <td>4.624622e-30</td>\n",
       "      <td>610.0</td>\n",
       "      <td>361.0</td>\n",
       "      <td>260.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>heterocycle biosynthetic process</th>\n",
       "      <td>2.715556</td>\n",
       "      <td>3.270897e-15</td>\n",
       "      <td>272.0</td>\n",
       "      <td>166.0</td>\n",
       "      <td>123.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>organic cyclic compound biosynthetic process</th>\n",
       "      <td>2.689241</td>\n",
       "      <td>6.209807e-15</td>\n",
       "      <td>273.0</td>\n",
       "      <td>166.0</td>\n",
       "      <td>123.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>organic substance transport</th>\n",
       "      <td>2.660079</td>\n",
       "      <td>7.593590e-04</td>\n",
       "      <td>53.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>29.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>purine nucleoside triphosphate biosynthetic process</th>\n",
       "      <td>2.635795</td>\n",
       "      <td>1.616744e-03</td>\n",
       "      <td>45.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>26.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nucleoside triphosphate biosynthetic process</th>\n",
       "      <td>2.635795</td>\n",
       "      <td>1.616744e-03</td>\n",
       "      <td>45.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>26.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>carbohydrate transport</th>\n",
       "      <td>2.381110</td>\n",
       "      <td>7.704527e-03</td>\n",
       "      <td>40.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>23.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>organic substance biosynthetic process</th>\n",
       "      <td>2.264725</td>\n",
       "      <td>4.992966e-24</td>\n",
       "      <td>740.0</td>\n",
       "      <td>405.0</td>\n",
       "      <td>311.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>electron transport chain</th>\n",
       "      <td>2.216470</td>\n",
       "      <td>2.344430e-03</td>\n",
       "      <td>64.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>34.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>macromolecule metabolic process</th>\n",
       "      <td>2.151585</td>\n",
       "      <td>1.115028e-20</td>\n",
       "      <td>714.0</td>\n",
       "      <td>384.0</td>\n",
       "      <td>301.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coenzyme metabolic process</th>\n",
       "      <td>1.993809</td>\n",
       "      <td>4.310858e-10</td>\n",
       "      <td>364.0</td>\n",
       "      <td>195.0</td>\n",
       "      <td>161.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cofactor metabolic process</th>\n",
       "      <td>1.853103</td>\n",
       "      <td>1.152495e-08</td>\n",
       "      <td>382.0</td>\n",
       "      <td>198.0</td>\n",
       "      <td>168.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein metabolic process</th>\n",
       "      <td>1.789623</td>\n",
       "      <td>1.197463e-07</td>\n",
       "      <td>368.0</td>\n",
       "      <td>188.0</td>\n",
       "      <td>162.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nucleobase-containing compound metabolic process</th>\n",
       "      <td>1.764399</td>\n",
       "      <td>6.148139e-11</td>\n",
       "      <td>623.0</td>\n",
       "      <td>311.0</td>\n",
       "      <td>265.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>heterocycle metabolic process</th>\n",
       "      <td>1.732269</td>\n",
       "      <td>8.344271e-11</td>\n",
       "      <td>664.0</td>\n",
       "      <td>328.0</td>\n",
       "      <td>281.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>organic cyclic compound metabolic process</th>\n",
       "      <td>1.720717</td>\n",
       "      <td>1.531958e-10</td>\n",
       "      <td>666.0</td>\n",
       "      <td>328.0</td>\n",
       "      <td>282.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cellular aromatic compound metabolic process</th>\n",
       "      <td>1.707894</td>\n",
       "      <td>2.602582e-10</td>\n",
       "      <td>664.0</td>\n",
       "      <td>326.0</td>\n",
       "      <td>281.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nitrogen compound metabolic process</th>\n",
       "      <td>1.486787</td>\n",
       "      <td>5.816754e-09</td>\n",
       "      <td>1198.0</td>\n",
       "      <td>539.0</td>\n",
       "      <td>491.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>purine nucleoside triphosphate metabolic process</th>\n",
       "      <td>1.474152</td>\n",
       "      <td>1.077848e-03</td>\n",
       "      <td>312.0</td>\n",
       "      <td>146.0</td>\n",
       "      <td>140.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nucleoside triphosphate metabolic process</th>\n",
       "      <td>1.474152</td>\n",
       "      <td>1.077848e-03</td>\n",
       "      <td>312.0</td>\n",
       "      <td>146.0</td>\n",
       "      <td>140.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pyridine-containing compound metabolic process</th>\n",
       "      <td>1.465692</td>\n",
       "      <td>3.960812e-03</td>\n",
       "      <td>252.0</td>\n",
       "      <td>118.0</td>\n",
       "      <td>115.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nicotinamide nucleotide metabolic process</th>\n",
       "      <td>1.465692</td>\n",
       "      <td>3.960812e-03</td>\n",
       "      <td>252.0</td>\n",
       "      <td>118.0</td>\n",
       "      <td>115.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oxidoreduction coenzyme metabolic process</th>\n",
       "      <td>1.465692</td>\n",
       "      <td>3.960812e-03</td>\n",
       "      <td>252.0</td>\n",
       "      <td>118.0</td>\n",
       "      <td>115.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATP metabolic process</th>\n",
       "      <td>1.463108</td>\n",
       "      <td>1.322288e-03</td>\n",
       "      <td>311.0</td>\n",
       "      <td>145.0</td>\n",
       "      <td>139.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nucleoside diphosphate metabolic process</th>\n",
       "      <td>1.461769</td>\n",
       "      <td>4.572463e-03</td>\n",
       "      <td>248.0</td>\n",
       "      <td>116.0</td>\n",
       "      <td>113.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nucleotide phosphorylation</th>\n",
       "      <td>1.461769</td>\n",
       "      <td>4.572463e-03</td>\n",
       "      <td>248.0</td>\n",
       "      <td>116.0</td>\n",
       "      <td>113.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>purine nucleoside monophosphate metabolic process</th>\n",
       "      <td>1.439221</td>\n",
       "      <td>2.195690e-03</td>\n",
       "      <td>318.0</td>\n",
       "      <td>147.0</td>\n",
       "      <td>142.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ribonucleoside metabolic process</th>\n",
       "      <td>1.437421</td>\n",
       "      <td>2.148045e-03</td>\n",
       "      <td>316.0</td>\n",
       "      <td>146.0</td>\n",
       "      <td>141.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>purine nucleoside metabolic process</th>\n",
       "      <td>1.437421</td>\n",
       "      <td>2.148045e-03</td>\n",
       "      <td>316.0</td>\n",
       "      <td>146.0</td>\n",
       "      <td>141.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nucleoside monophosphate metabolic process</th>\n",
       "      <td>1.421570</td>\n",
       "      <td>2.781896e-03</td>\n",
       "      <td>320.0</td>\n",
       "      <td>147.0</td>\n",
       "      <td>143.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>generation of precursor metabolites and energy</th>\n",
       "      <td>1.420085</td>\n",
       "      <td>1.417035e-03</td>\n",
       "      <td>369.0</td>\n",
       "      <td>169.0</td>\n",
       "      <td>163.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ribose phosphate metabolic process</th>\n",
       "      <td>1.414790</td>\n",
       "      <td>2.925167e-03</td>\n",
       "      <td>323.0</td>\n",
       "      <td>148.0</td>\n",
       "      <td>144.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>purine ribonucleotide metabolic process</th>\n",
       "      <td>1.412897</td>\n",
       "      <td>3.410280e-03</td>\n",
       "      <td>321.0</td>\n",
       "      <td>147.0</td>\n",
       "      <td>143.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ribonucleotide metabolic process</th>\n",
       "      <td>1.412897</td>\n",
       "      <td>3.410280e-03</td>\n",
       "      <td>321.0</td>\n",
       "      <td>147.0</td>\n",
       "      <td>143.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>glycosyl compound metabolic process</th>\n",
       "      <td>1.395846</td>\n",
       "      <td>4.283878e-03</td>\n",
       "      <td>323.0</td>\n",
       "      <td>147.0</td>\n",
       "      <td>144.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>carbohydrate derivative metabolic process</th>\n",
       "      <td>1.353845</td>\n",
       "      <td>6.088970e-03</td>\n",
       "      <td>369.0</td>\n",
       "      <td>165.0</td>\n",
       "      <td>163.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>64 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    odds_ratio       p_value  \\\n",
       "movement of cell or subcellular component            99.007941  5.477514e-46   \n",
       "locomotion                                           99.007941  5.477514e-46   \n",
       "S-adenosylmethionine metabolic process                7.111075  1.122956e-15   \n",
       "RNA biosynthetic process                              6.894115  2.009358e-25   \n",
       "nucleic acid-templated transcription                  6.766247  8.142000e-25   \n",
       "sulfur compound biosynthetic process                  6.690520  2.883961e-15   \n",
       "amide biosynthetic process                            6.037490  3.428364e-29   \n",
       "peptide metabolic process                             5.895077  1.094698e-28   \n",
       "cellular amide metabolic process                      5.846131  5.978251e-29   \n",
       "macromolecule biosynthetic process                    4.639342  1.728650e-48   \n",
       "cellular macromolecule biosynthetic process           4.602911  6.335099e-48   \n",
       "coenzyme biosynthetic process                         4.270630  3.248741e-12   \n",
       "cellular nitrogen compound biosynthetic process       3.950762  1.881661e-42   \n",
       "regulation of gene expression                         3.858625  4.460849e-03   \n",
       "nucleobase-containing compound biosynthetic pro...    3.323436  1.538432e-18   \n",
       "sulfur compound metabolic process                     3.202878  3.417680e-09   \n",
       "phosphoenolpyruvate-dependent sugar phosphotran...    3.179097  1.243402e-03   \n",
       "cellular biosynthetic process                         2.990730  1.842103e-41   \n",
       "cellular protein metabolic process                    2.934089  5.018869e-17   \n",
       "RNA metabolic process                                 2.926322  5.382106e-15   \n",
       "cofactor biosynthetic process                         2.880466  2.447821e-08   \n",
       "energy coupled proton transport, down electroch...    2.876926  1.012332e-03   \n",
       "aromatic compound biosynthetic process                2.788336  1.540568e-15   \n",
       "nucleic acid metabolic process                        2.719766  9.230484e-14   \n",
       "cellular macromolecule metabolic process              2.718431  4.624622e-30   \n",
       "heterocycle biosynthetic process                      2.715556  3.270897e-15   \n",
       "organic cyclic compound biosynthetic process          2.689241  6.209807e-15   \n",
       "organic substance transport                           2.660079  7.593590e-04   \n",
       "purine nucleoside triphosphate biosynthetic pro...    2.635795  1.616744e-03   \n",
       "nucleoside triphosphate biosynthetic process          2.635795  1.616744e-03   \n",
       "...                                                        ...           ...   \n",
       "carbohydrate transport                                2.381110  7.704527e-03   \n",
       "organic substance biosynthetic process                2.264725  4.992966e-24   \n",
       "electron transport chain                              2.216470  2.344430e-03   \n",
       "macromolecule metabolic process                       2.151585  1.115028e-20   \n",
       "coenzyme metabolic process                            1.993809  4.310858e-10   \n",
       "cofactor metabolic process                            1.853103  1.152495e-08   \n",
       "protein metabolic process                             1.789623  1.197463e-07   \n",
       "nucleobase-containing compound metabolic process      1.764399  6.148139e-11   \n",
       "heterocycle metabolic process                         1.732269  8.344271e-11   \n",
       "organic cyclic compound metabolic process             1.720717  1.531958e-10   \n",
       "cellular aromatic compound metabolic process          1.707894  2.602582e-10   \n",
       "nitrogen compound metabolic process                   1.486787  5.816754e-09   \n",
       "purine nucleoside triphosphate metabolic process      1.474152  1.077848e-03   \n",
       "nucleoside triphosphate metabolic process             1.474152  1.077848e-03   \n",
       "pyridine-containing compound metabolic process        1.465692  3.960812e-03   \n",
       "nicotinamide nucleotide metabolic process             1.465692  3.960812e-03   \n",
       "oxidoreduction coenzyme metabolic process             1.465692  3.960812e-03   \n",
       "ATP metabolic process                                 1.463108  1.322288e-03   \n",
       "nucleoside diphosphate metabolic process              1.461769  4.572463e-03   \n",
       "nucleotide phosphorylation                            1.461769  4.572463e-03   \n",
       "purine nucleoside monophosphate metabolic process     1.439221  2.195690e-03   \n",
       "ribonucleoside metabolic process                      1.437421  2.148045e-03   \n",
       "purine nucleoside metabolic process                   1.437421  2.148045e-03   \n",
       "nucleoside monophosphate metabolic process            1.421570  2.781896e-03   \n",
       "generation of precursor metabolites and energy        1.420085  1.417035e-03   \n",
       "ribose phosphate metabolic process                    1.414790  2.925167e-03   \n",
       "purine ribonucleotide metabolic process               1.412897  3.410280e-03   \n",
       "ribonucleotide metabolic process                      1.412897  3.410280e-03   \n",
       "glycosyl compound metabolic process                   1.395846  4.283878e-03   \n",
       "carbohydrate derivative metabolic process             1.353845  6.088970e-03   \n",
       "\n",
       "                                                    total_annot  in_group  \\\n",
       "movement of cell or subcellular component                 116.0     114.0   \n",
       "locomotion                                                116.0     114.0   \n",
       "S-adenosylmethionine metabolic process                     84.0      68.0   \n",
       "RNA biosynthetic process                                  145.0     116.0   \n",
       "nucleic acid-templated transcription                      143.0     114.0   \n",
       "sulfur compound biosynthetic process                       85.0      68.0   \n",
       "amide biosynthetic process                                187.0     145.0   \n",
       "peptide metabolic process                                 188.0     145.0   \n",
       "cellular amide metabolic process                          191.0     147.0   \n",
       "macromolecule biosynthetic process                        432.0     307.0   \n",
       "cellular macromolecule biosynthetic process               430.0     305.0   \n",
       "coenzyme biosynthetic process                             103.0      74.0   \n",
       "cellular nitrogen compound biosynthetic process           462.0     313.0   \n",
       "regulation of gene expression                              21.0      15.0   \n",
       "nucleobase-containing compound biosynthetic pro...        236.0     155.0   \n",
       "sulfur compound metabolic process                         111.0      73.0   \n",
       "phosphoenolpyruvate-dependent sugar phosphotran...         36.0      24.0   \n",
       "cellular biosynthetic process                             742.0     448.0   \n",
       "cellular protein metabolic process                        266.0     167.0   \n",
       "RNA metabolic process                                     232.0     146.0   \n",
       "cofactor biosynthetic process                             120.0      76.0   \n",
       "energy coupled proton transport, down electroch...         42.0      27.0   \n",
       "aromatic compound biosynthetic process                    266.0     164.0   \n",
       "nucleic acid metabolic process                            240.0     147.0   \n",
       "cellular macromolecule metabolic process                  610.0     361.0   \n",
       "heterocycle biosynthetic process                          272.0     166.0   \n",
       "organic cyclic compound biosynthetic process              273.0     166.0   \n",
       "organic substance transport                                53.0      33.0   \n",
       "purine nucleoside triphosphate biosynthetic pro...         45.0      28.0   \n",
       "nucleoside triphosphate biosynthetic process               45.0      28.0   \n",
       "...                                                         ...       ...   \n",
       "carbohydrate transport                                     40.0      24.0   \n",
       "organic substance biosynthetic process                    740.0     405.0   \n",
       "electron transport chain                                   64.0      37.0   \n",
       "macromolecule metabolic process                           714.0     384.0   \n",
       "coenzyme metabolic process                                364.0     195.0   \n",
       "cofactor metabolic process                                382.0     198.0   \n",
       "protein metabolic process                                 368.0     188.0   \n",
       "nucleobase-containing compound metabolic process          623.0     311.0   \n",
       "heterocycle metabolic process                             664.0     328.0   \n",
       "organic cyclic compound metabolic process                 666.0     328.0   \n",
       "cellular aromatic compound metabolic process              664.0     326.0   \n",
       "nitrogen compound metabolic process                      1198.0     539.0   \n",
       "purine nucleoside triphosphate metabolic process          312.0     146.0   \n",
       "nucleoside triphosphate metabolic process                 312.0     146.0   \n",
       "pyridine-containing compound metabolic process            252.0     118.0   \n",
       "nicotinamide nucleotide metabolic process                 252.0     118.0   \n",
       "oxidoreduction coenzyme metabolic process                 252.0     118.0   \n",
       "ATP metabolic process                                     311.0     145.0   \n",
       "nucleoside diphosphate metabolic process                  248.0     116.0   \n",
       "nucleotide phosphorylation                                248.0     116.0   \n",
       "purine nucleoside monophosphate metabolic process         318.0     147.0   \n",
       "ribonucleoside metabolic process                          316.0     146.0   \n",
       "purine nucleoside metabolic process                       316.0     146.0   \n",
       "nucleoside monophosphate metabolic process                320.0     147.0   \n",
       "generation of precursor metabolites and energy            369.0     169.0   \n",
       "ribose phosphate metabolic process                        323.0     148.0   \n",
       "purine ribonucleotide metabolic process                   321.0     147.0   \n",
       "ribonucleotide metabolic process                          321.0     147.0   \n",
       "glycosyl compound metabolic process                       323.0     147.0   \n",
       "carbohydrate derivative metabolic process                 369.0     165.0   \n",
       "\n",
       "                                                    expected  \n",
       "movement of cell or subcellular component               57.0  \n",
       "locomotion                                              57.0  \n",
       "S-adenosylmethionine metabolic process                  43.0  \n",
       "RNA biosynthetic process                                70.0  \n",
       "nucleic acid-templated transcription                    69.0  \n",
       "sulfur compound biosynthetic process                    44.0  \n",
       "amide biosynthetic process                              88.0  \n",
       "peptide metabolic process                               88.0  \n",
       "cellular amide metabolic process                        89.0  \n",
       "macromolecule biosynthetic process                     188.0  \n",
       "cellular macromolecule biosynthetic process            188.0  \n",
       "coenzyme biosynthetic process                           52.0  \n",
       "cellular nitrogen compound biosynthetic process        200.0  \n",
       "regulation of gene expression                           14.0  \n",
       "nucleobase-containing compound biosynthetic pro...     108.0  \n",
       "sulfur compound metabolic process                       55.0  \n",
       "phosphoenolpyruvate-dependent sugar phosphotran...      21.0  \n",
       "cellular biosynthetic process                          312.0  \n",
       "cellular protein metabolic process                     121.0  \n",
       "RNA metabolic process                                  106.0  \n",
       "cofactor biosynthetic process                           59.0  \n",
       "energy coupled proton transport, down electroch...      24.0  \n",
       "aromatic compound biosynthetic process                 121.0  \n",
       "nucleic acid metabolic process                         110.0  \n",
       "cellular macromolecule metabolic process               260.0  \n",
       "heterocycle biosynthetic process                       123.0  \n",
       "organic cyclic compound biosynthetic process           123.0  \n",
       "organic substance transport                             29.0  \n",
       "purine nucleoside triphosphate biosynthetic pro...      26.0  \n",
       "nucleoside triphosphate biosynthetic process            26.0  \n",
       "...                                                      ...  \n",
       "carbohydrate transport                                  23.0  \n",
       "organic substance biosynthetic process                 311.0  \n",
       "electron transport chain                                34.0  \n",
       "macromolecule metabolic process                        301.0  \n",
       "coenzyme metabolic process                             161.0  \n",
       "cofactor metabolic process                             168.0  \n",
       "protein metabolic process                              162.0  \n",
       "nucleobase-containing compound metabolic process       265.0  \n",
       "heterocycle metabolic process                          281.0  \n",
       "organic cyclic compound metabolic process              282.0  \n",
       "cellular aromatic compound metabolic process           281.0  \n",
       "nitrogen compound metabolic process                    491.0  \n",
       "purine nucleoside triphosphate metabolic process       140.0  \n",
       "nucleoside triphosphate metabolic process              140.0  \n",
       "pyridine-containing compound metabolic process         115.0  \n",
       "nicotinamide nucleotide metabolic process              115.0  \n",
       "oxidoreduction coenzyme metabolic process              115.0  \n",
       "ATP metabolic process                                  139.0  \n",
       "nucleoside diphosphate metabolic process               113.0  \n",
       "nucleotide phosphorylation                             113.0  \n",
       "purine nucleoside monophosphate metabolic process      142.0  \n",
       "ribonucleoside metabolic process                       141.0  \n",
       "purine nucleoside metabolic process                    141.0  \n",
       "nucleoside monophosphate metabolic process             143.0  \n",
       "generation of precursor metabolites and energy         163.0  \n",
       "ribose phosphate metabolic process                     144.0  \n",
       "purine ribonucleotide metabolic process                143.0  \n",
       "ribonucleotide metabolic process                       143.0  \n",
       "glycosyl compound metabolic process                    144.0  \n",
       "carbohydrate derivative metabolic process              163.0  \n",
       "\n",
       "[64 rows x 5 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(enrichment_df('RT_Unenriched', 'RT_Enriched', bp_map)\n",
    " .query('odds_ratio > 1 and p_value < 0.01')\n",
    " .sort_values('odds_ratio', ascending=False))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## RAG - Enriched vs Unenriched - MF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "(enrichment_df('RAG_Enriched', 'RAG_Unenriched', mf_map)\n",
    " .query('odds_ratio > 1 and p_value < 0.01')\n",
    " .sort_values('odds_ratio', ascending=False))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## RAG - Enriched vs Unenriched - BP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "(enrichment_df('RAG_Enriched', 'RAG_Unenriched', bp_map)\n",
    " .query('odds_ratio > 1 and p_value < 0.01')\n",
    " .sort_values('odds_ratio', ascending=False))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## RAG - Unenriched vs Enriched - MF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "(enrichment_df('RAG_Unenriched', 'RAG_Enriched', mf_map)\n",
    " .query('odds_ratio > 1 and p_value < 0.01')\n",
    " .sort_values('odds_ratio', ascending=False))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## RAG - Unenriched vs Enriched - BP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "(enrichment_df('RAG_Unenriched', 'RAG_Enriched', bp_map)\n",
    " .query('odds_ratio > 1 and p_value < 0.01')\n",
    " .sort_values('odds_ratio', ascending=False))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
